source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -0,0 +1,77 @@
+#include "source/extensions/filters/http/header_mutation/header_mutation.h"
+
+#include <cstdint>
+#include <memory>
+
+#include "source/common/config/utility.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/common/http/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HeaderMutation {
+
+void Mutations::mutateRequestHeaders(Http::RequestHeaderMap& request_headers,
+                                     const StreamInfo::StreamInfo& stream_info) const {
+  request_mutations_.evaluateHeaders(request_headers, request_headers,
+                                     *Http::StaticEmptyHeaders::get().response_headers,
+                                     stream_info);
+}
+
+void Mutations::mutateResponseHeaders(const Http::RequestHeaderMap& request_headers,
+                                      Http::ResponseHeaderMap& response_headers,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  response_mutations_.evaluateHeaders(response_headers, request_headers, response_headers,
+                                      stream_info);
+}
+
+PerRouteHeaderMutation::PerRouteHeaderMutation(const PerRouteProtoConfig& config)
+    : mutations_(config.mutations()) {}
+
+HeaderMutationConfig::HeaderMutationConfig(const ProtoConfig& config)
+    : mutations_(config.mutations()) {}
+
+Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  config_->mutations().mutateRequestHeaders(headers, decoder_callbacks_->streamInfo());
```

Comment:
Do you want to do both per-route and filter-level config, or have route-specific override the filter config?

Reply:
Maybe call this out in docs / comments somewhere.

Reply:
I think I has called this out in the proto API.

Reply:
Yeah, I made a small suggestion to make it clearer but you are right.